flink-cdc-connect/flink-cdc-source-connectors/flink-cdc-base/src/main/java/org/apache/flink/cdc/connectors/base/source/metrics/SourceEnumeratorMetrics.java

```diff
@@ -0,0 +1,251 @@
+    private volatile int numTablesRemaining = 0;
+
+    // Map for managing per-table metrics by table identifier
+    // Key: Identifier of the table
+    // Value: TableMetrics related to the table
+    private final Map<TableId, TableMetrics> tableMetricsMap = new ConcurrentHashMap<>();
+
+    public SourceEnumeratorMetrics(SplitEnumeratorMetricGroup metricGroup) {
+        this.metricGroup = metricGroup;
+        metricGroup.gauge(IS_SNAPSHOTTING, () -> isSnapshotting);
+        metricGroup.gauge(IS_STREAM_READING, () -> isStreamReading);
+        metricGroup.gauge(NUM_TABLES_REMAINING, () -> numTablesRemaining);
+    }
+
+    public void enterSnapshotPhase() {
+        this.isSnapshotting = 1;
+    }
+
+    public void exitSnapshotPhase() {
+        this.isSnapshotting = 0;
+    }
+
+    public void enterStreamReading() {
+        this.isStreamReading = 1;
+    }
+
+    public void exitStreamReading() {
+        this.isStreamReading = 0;
+    }
+
+    public void registerMetrics(
+            Gauge<Integer> numTablesSnapshotted,
+            Gauge<Integer> numSnapshotSplitsProcessed,
+            Gauge<Integer> numSnapshotSplitsRemaining) {
+        metricGroup.gauge(NUM_TABLES_SNAPSHOTTED, numTablesSnapshotted);
+        metricGroup.gauge(NUM_SNAPSHOT_SPLITS_PROCESSED, numSnapshotSplitsProcessed);
+        metricGroup.gauge(NUM_SNAPSHOT_SPLITS_REMAINING, numSnapshotSplitsRemaining);
+    }
+
+    public void addNewTables(int numNewTables) {
+        numTablesRemaining += numNewTables;
+    }
+
+    public void startSnapshotTables(int numSnapshottedTables) {
+        numTablesRemaining -= numSnapshottedTables;
+    }
+
+    public TableMetrics getTableMetrics(TableId tableId) {
+        return tableMetricsMap.computeIfAbsent(
+                tableId,
+                key -> new TableMetrics(key.catalog(), key.schema(), key.table(), metricGroup));
+    }
+
+    // ----------------------------------- Helper classes --------------------------------
+
+    /**
+     * Collection class for managing metrics of a table.
+     *
+     * <p>Metrics of table level are registered in its corresponding subgroup under the {@link
+     * SplitEnumeratorMetricGroup}.
+     */
+    public static class TableMetrics {
+        private AtomicInteger numSnapshotSplitsProcessed = new AtomicInteger(0);
+        private AtomicInteger numSnapshotSplitsRemaining = new AtomicInteger(0);
+        private AtomicInteger numSnapshotSplitsFinished = new AtomicInteger(0);
+        private volatile long snapshotStartTime = UNDEFINED;
+        private volatile long snapshotEndTime = UNDEFINED;
+
+        private Set<String> remainingSplitIds =
+                Collections.newSetFromMap(new ConcurrentHashMap<>());
+        private Set<String> processedSplitIds =
+                Collections.newSetFromMap(new ConcurrentHashMap<>());
+        private Set<String> finishedSplitIds = Collections.newSetFromMap(new ConcurrentHashMap<>());
```

Comment:
I think we don't need these fields.
- There is no metric registered about `remainingSplitIds` and `processedSplitIds`
- `finishedSplitIds` could be stored in the assigner instead of this metric class

Reply:
These metrics (`numSnapshotSplitsProcessed`/`numSnapshotSplitsRemaining`/`numSnapshotSplitsFinished`) can be used to calculate the synchronization progress of a specified table.(`remainingSplitIds`/`processedSplitIds`/`finishedSplitIds`) are used for deduplication